File: Easy/Binary_Tree_Level_Order_Traversal/binaryTreeLevelOrderTraversal.js
// Given a binary tree, return the bottom-up level order traversal of its nodes' values. (ie, from left to right, level by level from leaf to root).

// For example:
// Given binary tree [3,9,20,null,null,15,7],
//     3
//    / \
//   9  20
//     /  \
//    15   7
// return its bottom-up level order traversal as:
// [
//   [15,7],
//   [9,20],
//   [3]
// ]

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrderBottom = function(root) {
  if(!root) {
    return [];
  }
  
  const queue = [root];
  const bottomUp = [];
  
  while(queue.length > 0) { 
    let currLevel = [];
    let queueLen = queue.length;
    
    for(let i = 0; i < queueLen; i++) { 
      const popped = queue.shift();
      
      if(popped.left) {
        queue.push(popped.left);  
      }
      if(popped.right) {
        queue.push(popped.right);
      }      
      
      currLevel.push(popped.val);
    }
    
    bottomUp.unshift(currLevel);    
  }
  
  return bottomUp;
};

File: Medium/Binary_Tree_Level_Order_Traversal/binaryTreeLevelOrderTraversal.cs
// Given a binary tree, return the level order traversal of its nodes' values. (ie, from left to right, level by level).

// For example:
// Given binary tree [3,9,20,null,null,15,7],
//     3
//    / \
//   9  20
//     /  \
//    15   7

// return its level order traversal as:
// [
//   [3],
//   [9,20],
//   [15,7]
// ]

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
  public IList<IList<int>> LevelOrder(TreeNode root) {
    var result = new List<IList<int>>();
    if(root == null) return result;
    
    var q = new Queue<TreeNode>();
    q.Enqueue(root);
    
    while(q.Count > 0) {
      var count = q.Count;
      var curLevel = new List<int>();
      while(count-- > 0) {
        if(q.Peek().left != null) q.Enqueue(q.Peek().left);
        if(q.Peek().right != null) q.Enqueue(q.Peek().right);
        curLevel.Add(q.Dequeue().val);
      }
      result.Add(curLevel);
    }
    return result;
  }  
}